src/clj/puppetlabs/puppetserver/jruby_request.clj

```diff
@@ -67,10 +67,14 @@
      (handler (assoc request :jruby-instance jruby-instance)))))
 
 (defn get-environment-from-request
-  "Gets the environment from a request."
+  "Gets the environment from the URL or query string of a request."
   [req]
-  (-> req
-      (get-in [:params "environment"])))
+  ;; :environment is the value of a path component, "environment" is the value
+  ;; of a query string parameter. If environment is an inherent part of the
+  ;; resource being requested (part of the path), we prefer that so it doesn't
+  ;; get overridden by a rogue query string parameter.
+  (or (get-in req [:params :environment])
+      (get-in req [:params "environment"])))
```

Comment:
@nicklewis would you mind making that one minor change for now, then, and we can get this merged and spare you the rest of our musings about disparate sources for request parameters? :)